docs/design-documents/20221028-dead-letter-queue.md

```diff
@@ -0,0 +1,144 @@
+# Dead letter queue
+
+## Introduction
+
+In Conduit the concept of a Dead letter queue (DLQ) describes the place where
+messages are sent if they are negatively acknowledged (nacked). This can happen
+for a number of reasons:
+
+- A node in the pipeline experiences an error while processing a message (e.g.
+  processor).
+- The destination connector fails to process a message and sends back a nack.
+- The pipeline is stopped forcefully causing in-flight messages to be nacked.
+
+### Goals
+
+The purpose of a DLQ is to make a pipeline resilient to errors that might
+happen while processing messages.
+
+The design for the DLQ should support the following:
+
+- Allow user to reroute nacked messages to any Conduit connector (builtin or
+  standalone).
+- A message sent to the DLQ should reflect the contents it had just before the
+  error happened. Example: if a message went through 2 processors and failed
+  while being processed by a 3rd processor, the DLQ message should reflect any
+  transforms done by the first 2 processors.
+- Ensure message order is preserved when messages are sent to the DLQ.
+  In other words - the order in which messages were produced by the source
+  connector should be preserved, even if messages are nacked in a different
+  order.
+- If a message failed to be sent to the DLQ the nack action should fail and
+  stop the pipeline.
```

Comment:
we could add another layer here, maybe create a simple file destination and write the failed records there, with a message telling the user what happened. Of course, this could fail too and then we stop the pipelines. WDYT?

Reply:
oh just noticed this in the future work, it's the same as a fallback DLQ but we decide what the destination will be, not the user.

Reply:
Right, that would be essentially the same as having a fallback to `builtin:file`. If we want this feature I'd rather go with the file connector than write one in Conduit for this purpose.

Reply:
Yes I meant using the builtin connector! when I mentioned

> simple file destination

I meant creating the actual output file that would have the records.

Reply:
I'm thinking what we're trying to achieve with this? Sounds like we are trying to make sure that the pipeline can keep on running even if a message fails to be pushed to the DLQ. In this case I would rather provide a generic solution for chaining DLQs as fallbacks, I wouldn't want to make the assumption that the user is fine with writing records to a file.

That said, I think even without the possibility of fallbacks we should already cover most of our users needs, so I wrote this down as future work.